db/compaction/compaction.cc

```diff
@@ -383,7 +383,14 @@ bool Compaction::KeyNotExistsBeyondOutputLevel(
         auto* f = files[level_ptrs->at(lvl)];
         if (user_cmp->Compare(user_key, f->largest.user_key()) <= 0) {
           // We've advanced far enough
-          if (user_cmp->Compare(user_key, f->smallest.user_key()) >= 0) {
+          // In the presence of user-defined timestamp, we may need to handle
+          // the case in which f->smallest.user_key() (including ts) has the
+          // same user key, but the ts part is smaller. If so,
+          // Compare(user_key, f->smallest.user_key()) returns -1.
+          // That's why we need CompareWithoutTimestamp().
+          if (user_cmp->Compare(user_key, f->smallest.user_key()) >= 0 ||
+              user_cmp->CompareWithoutTimestamp(user_key,
```

Comment:
Good point. I think a single `CompareWithoutTimestamp()` should do.
If `f->smallest.user_key` is a really different user key (excluding timestamp), then `CompareWithoutTimestamp()` will return 1, which is the same as before.
If `f->smallest.user_key` is actually the same user key, then it will return 0.